lua/checks.lua

```diff
@@ -0,0 +1,192 @@
+require("lua.compiler")
+require("lua.util")
+
+function CheckSystemFamily()
+	Check("Determining operating system family")
+	local path = os.getenv("PATH")
+
+	if not path then
+		Error()
+	end
+
+	-- Try find part of a absolute drive path like 'C:\Windows\System32'
+	dos = string.match(path, ":\\")
```

Comment:
Is this 100% safe? Should perhaps run `uname -o` first to check for Posix?

Reply:
This is 100% safe and works 99.99% of the time the 0.01% being circumstances where DOS is too old or Unix/Windows is crippled beyond the point of being usable. Never the less I'll think about how I can harden this check and error out under any ambigious circumstances.

The check works by reading the path enviroment value. On Windows 95+ it will always contain at least one absolute directory that will match `:\`. The path seperator on DOS/CMD is `;` compared to `:` on Unix:
Example from DOSBox-X:
```
>echo %PATH%
Z:\;Z:\SYSTEM;Z:\BIN;Z:\DOS;Z:\4DOS;Z:\DEBUG;Z:\TEXTUTIL;C:\DOS\BIN;C:\DOS\DEVEL\LUA\BIN
```
Example from Wine: 
```
>echo %PATH%
C:\windows\system32;C:\windows;C:\windows\system32\WindowsPowershell\v1.0;
```
Example from Linux 
```
$ echo $PATH
/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/bin:
```

> Should perhaps run uname -o first to check for Posix?

Unfortunatly a call like this from `COMMAND.COM`/`CMD.EXE` could be made accidentally unreliable if `uname.exe` does exist in `%PATH%` (a MSYS or Cygwin installation for example). Additionally from a security standpoint it's trivial to hijack a call from these shells, you can try the following yourself:
```
C:\>ver
Microsoft Windows 10.0.19043

C:\>where ping
C:\windows\system32\ping.exe

C:\>ping /?
Usage: ping [-n count] [-w timeout] [-l buffer_length] target_name

Options:
    -n  Number of echo requests to send.
    -w  Timeout in milliseconds to wait for each reply.
    -l  Length of send buffer.

C:\>echo @ECHO OFF > ping.bat

C:\>echo echo DOS GOT DOS-d >> ping.bat

C:\>where ping
C:\ping.bat
C:\windows\system32\ping.exe

C:\>ping /?
DOS GOT DOS-d

C:\>
```
The only exception to this flaw are in-built commands
```
C:\>md /?
MD <name> is the short version of MKDIR. It creates a subdirectory.

C:\>where md
File not found

C:\>echo This won't work > md.bat

C:\>where md
C:\md.bat

C:\>md /?
MD <name> is the short version of MKDIR. It creates a subdirectory.

C:\>
```

Reply:
This code has been rewritten. See replacement implementation: https://github.com/gvanem/Watt-32/blob/5debd9c81a1cb56e082d2ffa66279f2d3ccfae72/lua/boot.lua#L134-L168